File: lab7-vars/python/playlistVars.py
class Piece:
    def __init__(self, name, performer, length_secs): # constructor
        self.name = name
        self.performer = performer
        self.length_secs = length_secs
        self.check()

    def __repr__(self):  # formatting (like toString())
        result = "%s by %s (%.2fs)" % (self.name, self.performer, self.length_secs)
        return result

    def check(self):
        # do some dynamic type checking:
        assert self.name.__class__ == str
        assert self.performer.__class__ == str
        assert self.length_secs.__class__ == type(1.0)
        l = self.length_secs
        assert 0 < l and l < 36000

class Product:
    def __init__(self, name, brand):
        self.name = name
        self.brand = brand
        self.check()

    def __repr__(self):  # formatting (like toString())
        result = "%s by %s" % (self.name, self.brand)
        return result

    def check(self):
        # do some dynamic type checking:
        assert self.name.__class__ == str
        assert self.brand.__class__ == str

class Advert:
    def __init__(self, product, length_secs): # constructor
        self.product = product
        self.length_secs = length_secs
        self.check()

    def __repr__(self):  # formatting (like toString())
        result = "Advert for %s by %s (%.2fs)" % (self.product.name, self.product.brand, self.length_secs)
        return result

    def check(self):
        # do some dynamic type checking:
        assert self.product.__class__ == Product
        assert self.length_secs.__class__ == float
        l = self.length_secs
        assert 0 < l and l < 120

playlist1 = []

def definePlaylist1():
    piece1 = Piece("Moonlight", "C. Arrau", 17*60+26.0)
    piece2 = Piece("Pathetique", "D. Barenboim", 16*60+49.0)
    advert1 = Advert(Product("chocolate", "Yummm"), 15.0)
    playlist1 = [piece1, advert1, piece2]

definePlaylist1()

print("playlist1 = %s" % playlist1)

length1 = sum([ item.length_secs for item in playlist1 ])

print("lengths1 = %s" % length1)

File: lab7-vars/haskell/playlistVars.hs
module PlaylistVars where

import Text.Printf

data Product
    = Product
        {
          product_name :: String
        , product_brand :: String
        }
    deriving (Eq)

instance Show Product where 
    show (Product name brand) = 
        printf "%s by %s" name brand

data Item
    = Piece
        {
            item_name :: String,
            item_performer :: String,
            item_length_secs :: Float
        }
    | Advert
        {
            item_product :: Product,
            item_length_secs :: Float
        }
    deriving (Eq)

isAdvert (Advert _ _) = True
isAdvert _ = False

itemOK (Piece _ _ len) = 0 < len && len < 36000
itemOK (Advert _ len) = 0 < len && len < 120

instance (Show Item) where
    show (Piece name performer len) =
        printf "%s by %s (%.1fs)" name performer len
    show (Advert product len) =
        printf "Advert for %s (%.1fs)" (show product) len

playlist1 = [piece1, advert1, piece2]
    where
    piece1  = Piece "Moonlight" "C. Arrau"       (17*minutes+26*seconds)
    piece2  = Piece "Pathetique" "D. Barenboim"  (16*minutes+49*seconds)
    advert1 = Advert (Product "chocolate" "Yummm")          (15*seconds)
    minutes = 60*seconds -- TASK: identify the scope of variable "minutes"
    seconds = 1

length1 = sum [ item_length_secs item | item <- playlist1 ]
    -- OPTIONAL TASK: identify the scope of variable "item" above

main :: IO ()
main =
    do
    printf "playlist1 = %s\n" (show playlist1)
    printf "lenght1 = %s\n" (show length1)

